rmf_gazebo_plugins/src/readonly.cpp

```diff
@@ -0,0 +1,466 @@
+  for (std::size_t i =0; i < _graph.vertices.size(); ++i)
+    _wp_names.insert(std::make_pair(i, _graph.vertices[i].name));
+
+  _initialized_graph = true;
+
+}
+
+double ReadonlyPlugin::compute_ds(
+    const ignition::math::Pose3d& pose,
+    const std::size_t& wp)
+{
+  // TODO consider returning a nullptr instead
+  assert(_found_graph);
+  assert(wp < _graph.vertices.size());
+  
+  ignition::math::Vector3d world_position{pose.Pos().X(), pose.Pos().Y(), 0};
+  ignition::math::Vector3d graph_position{
+    _graph.vertices[wp].x,
+    _graph.vertices[wp].y,
+    0};
+
+  return std::abs((world_position - graph_position).Length());
+}
+
+void ReadonlyPlugin::initialize_start(const ignition::math::Pose3d& pose)
+{
+  if (_initialized_start)
+    return;
+
+  if (!_initialized_graph)
+    return;
+
+  bool found = false;
+  for (std::size_t i = 0; i < _graph.vertices.size(); i++)
+  {
+    if (_graph.vertices[i].name.c_str() == _start_wp_name)
+    {
+      found = true;
+      _start_wp = i;
+      RCLCPP_INFO(logger(), "Start waypoint found in nav graph");
+    }
+  }
+
+  // TODO find the closest wp if the coordiantes do not match
+  if (compute_ds(pose, _start_wp) < 1e-1)
+  {
+    _initialized_start = true;
+    // Here we initialzie the next waypoint
+    compute_path(pose);
+    RCLCPP_INFO(logger(), "Start waypoint successfully initialized");
+  }
+
+  else
+  {
+    RCLCPP_ERROR(
+      logger(),
+      "Spawn coordinates [%f,%f,%f] differs from that of waypoint [%s] in nav_graph [%f, %f, %f]",
+       pose.Pos().X(), pose.Pos().Y(),0,
+      _start_wp_name.c_str(),
+      _graph.vertices[_start_wp].x, _graph.vertices[_start_wp].y, 0);
+  }
+  
+}
+std::size_t ReadonlyPlugin::get_next_waypoint(const std::size_t& start_wp,
+    const ignition::math::Vector3d& heading,
+    const ignition::math::Pose3d& pose)
+{
+  // Return the waypoint closest to the robot in the direction of its heading
+  const auto& neighbors = _neighbor_map.find(start_wp)->second;
+
+  auto wp_it = neighbors.begin();
+  double max_dist = std::numeric_limits<double>::min();
+
+  for (auto it = neighbors.begin(); it != neighbors.end(); it++)
+  {
+    const auto& waypoint = _graph.vertices[*it];
+    ignition::math::Vector3d disp_vector{
+        waypoint.x - _graph.vertices[start_wp].x,
+        waypoint.y - _graph.vertices[start_wp].y,
+        0};
+    const double dist = heading.Dot(disp_vector.Normalize());
+    // Consider the waypoints with highest projected distance
+    if (dist > max_dist)
+    {
+      max_dist = dist;
+      wp_it = it;
+    }
+  }
+
+  return *wp_it;
+}
+
+ReadonlyPlugin::Path ReadonlyPlugin::compute_path(const ignition::math::Pose3d& pose)
+{
+  Path path;
+  path.resize(_lookahead);
+
+  std::size_t count = 0;
+  
+  auto start_wp = _start_wp;
+  const double current_yaw = pose.Rot().Euler().Z();
+  ignition::math::Vector3d heading{
+      std::cos(current_yaw), std::sin(current_yaw), 0.0};
+  
+  while (count < _lookahead)
```

Comment:
nitpick, looks like we can replace this with a for loop instead, then we can keep the scope of the variable `count` within the loop

Reply:
That's a good point.